LibWeb: Update stylesheet loading to use fetch infrastructure

srikavin opened the pull request:
This replaces the adhoc stylesheet loading implementation with one that more closely follows the spec, fixing some issues with stylesheet load events not firing.

MacDue commented:
Looks like you have some linter errors

(Just as an FYI you might want to setup the pre-commit to catch these before pushing: https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md#commit-hooks)

linusg commented on Userland/Libraries/LibWeb/Fetch/Body.cpp:
```diff
@@ -152,23 +146,42 @@ JS::NonnullGCPtr<JS::Promise> consume_body(JS::Realm& realm, BodyMixin const& ob
+    auto success_steps = [promise_capability, &vm, convertBytesToJSValue = move(convertBytesToJSValue)](ByteBuffer data) {
+        auto result = convertBytesToJSValue(move(data));
+        if (result.is_error()) {
+            // FIXME: run errorSteps with that exception.
```
Same here

linusg commented on Userland/Libraries/LibWeb/Fetch/Body.h:
```diff
@@ -41,7 +41,6 @@ class BodyMixin {
 
-[[nodiscard]] WebIDL::ExceptionOr<JS::Value> package_data(JS::Realm&, ByteBuffer, PackageDataType, Optional<MimeSniff::MimeType> const&);
-[[nodiscard]] JS::NonnullGCPtr<JS::Promise> consume_body(JS::Realm&, BodyMixin const&, PackageDataType);
+[[nodiscard]] JS::NonnullGCPtr<JS::Promise> consume_body(JS::Realm& realm, BodyMixin const& object, JS::SafeFunction<WebIDL::ExceptionOr<JS::Value>(AK::ByteBuffer)> convertBytesToJSValue);
```
Drop AK prefix, snake case

linusg commented on Userland/Libraries/LibWeb/Fetch/Infrastructure/HTTP/Bodies.cpp:
```diff
@@ -40,19 +41,35 @@ WebIDL::ExceptionOr<Body> Body::clone() const
-// https://fetch.spec.whatwg.org/#fully-reading-body-as-promise
-JS::NonnullGCPtr<JS::PromiseCapability> Body::fully_read_as_promise() const
+// https://fetch.spec.whatwg.org/#body-fully-read
+void Body::fully_read(JS::SafeFunction<void(ByteBuffer)> process_body, JS::SafeFunction<void()> process_body_error,
```
Use `TaskDestination` alias instead of typing out the variant again

linusg commented on Userland/Libraries/LibWeb/Fetch/Infrastructure/HTTP/Bodies.h:
```diff
@@ -31,7 +31,7 @@ class Body final {
     [[nodiscard]] WebIDL::ExceptionOr<Body> clone() const;
 
-    [[nodiscard]] JS::NonnullGCPtr<JS::PromiseCapability> fully_read_as_promise() const;
+    void fully_read(JS::SafeFunction<void(ByteBuffer)>, JS::SafeFunction<void()>, Variant<Empty, JS::NonnullGCPtr<JS::Object>>) const;
```
Likewise

linusg commented on Userland/Libraries/LibWeb/HTML/CORSSettingAttribute.cpp:
```diff
@@ -0,0 +1,28 @@
+namespace Web::HTML {
+
+// https://html.spec.whatwg.org/multipage/urls-and-fetching.html#cors-settings-attributes
+CORSSettingAttribute cors_setting_attribute_from_keyword(DeprecatedString keyword)
```
This can take a StringView

linusg commented on Userland/Libraries/LibWeb/HTML/CORSSettingAttribute.cpp:
```diff
@@ -0,0 +1,28 @@
+// https://html.spec.whatwg.org/multipage/urls-and-fetching.html#cors-settings-attributes
+CORSSettingAttribute cors_setting_attribute_from_keyword(DeprecatedString keyword)
+{
+    if (keyword.is_null()) {
```
Never use null strings in new code, use `Optional<T>` instead

linusg commented on Userland/Libraries/LibWeb/HTML/CORSSettingAttribute.cpp:
```diff
@@ -0,0 +1,28 @@
+    // The attribute's invalid value default is the Anonymous state
+    return CORSSettingAttribute::Anonymous;
+}
+}
```
Missing newline before closing namespace

linusg commented on Userland/Libraries/LibWeb/HTML/CORSSettingAttribute.h:
```diff
@@ -0,0 +1,22 @@
+    UseCredentials
+};
+
+CORSSettingAttribute cors_setting_attribute_from_keyword(DeprecatedString);
```
Make return value `[[nodiscard]]`

linusg commented on Userland/Libraries/LibWeb/HTML/PotentialCORSRequest.cpp:
```diff
@@ -0,0 +1,43 @@
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "PotentialCORSRequest.h"
```
Use `<LibWeb/*>` include style

linusg commented on Userland/Libraries/LibWeb/HTML/PotentialCORSRequest.cpp:
```diff
@@ -0,0 +1,43 @@
+
+    return request;
+}
+}
```
Missing newline before closing namespace

linusg commented on Userland/Libraries/LibWeb/HTML/PotentialCORSRequest.h:
```diff
@@ -0,0 +1,21 @@
+#include <LibWeb/Fetch/Infrastructure/HTTP/Requests.h>
+#include <LibWeb/HTML/CORSSettingAttribute.h>
+
+namespace Web::HTML {
```
Missing newline after opening namespace

linusg commented on Userland/Libraries/LibWeb/HTML/PotentialCORSRequest.h:
```diff
@@ -0,0 +1,21 @@
+};
+
+JS::NonnullGCPtr<Fetch::Infrastructure::Request> create_potential_CORS_request(JS::VM&, const AK::URL&, Optional<Fetch::Infrastructure::Request::Destination>, CORSSettingAttribute, SameOriginFallbackFlag = SameOriginFallbackFlag::No);
+}
```
Missing newline before closing namespace

linusg commented on Userland/Libraries/LibWeb/HTML/PotentialCORSRequest.h:
```diff
@@ -0,0 +1,21 @@
+    Yes,
+};
+
+JS::NonnullGCPtr<Fetch::Infrastructure::Request> create_potential_CORS_request(JS::VM&, const AK::URL&, Optional<Fetch::Infrastructure::Request::Destination>, CORSSettingAttribute, SameOriginFallbackFlag = SameOriginFallbackFlag::No);
```
nodiscard

linusg commented on Userland/Libraries/LibWeb/HTML/HTMLLinkElement.cpp:
```diff
@@ -177,7 +421,6 @@ void HTMLLinkElement::resource_did_load_favicon()
 
     document().check_favicon_after_loading_link_resource();
 }
-
```
Put this back :)

linusg commented on Userland/Libraries/LibWeb/HTML/HTMLLinkElement.cpp:
```diff
@@ -211,5 +454,4 @@ void HTMLLinkElement::visit_edges(Cell::Visitor& visitor)
     Base::visit_edges(visitor);
     visitor.visit(m_loaded_style_sheet);
 }
-
```
Put this back :)

srikavin commented on Userland/Libraries/LibWeb/Fetch/Body.cpp:
```diff
@@ -152,23 +146,42 @@ JS::NonnullGCPtr<JS::Promise> consume_body(JS::Realm& realm, BodyMixin const& ob
-        return package_data(realm, move(bytes), type, object.mime_type_impl());
+    // 3. Let errorSteps given error be to reject promise with error.
+    auto error_steps = [promise_capability, &vm]() {
+        // NOTE: The function body->fully_read expects no arguments for this callback,
+        //       even though the spec states this callback takes 'error'.
```
I might be misunderstanding the spec, but the spec's definition and use of `fully read` appear to be inconsistent.

The [definition for `fully read`](https://fetch.spec.whatwg.org/commit-snapshots/464326e8eb6a602122c030cd40042480a3c0e265/#body-fully-read) states `processBodyError must be an algorithm accepting no arguments`, and step 4 is `...If that threw an exception, then run errorSteps with that exception and return.` (and `errorSteps` isn't defined to take a parameter).

In this instance, processBodyError takes an error argument, but it's also used in [`fetch response handover`](https://fetch.spec.whatwg.org/commit-snapshots/464326e8eb6a602122c030cd40042480a3c0e265/#ref-for-fetch-params-process-response-consume-body%E2%91%A0) without the error parameter.

I guess we could just have `fully read` expect a `JS::Value exception` argument in the error callback, and have it be unused in the callback created in fetch response handover.

linusg replied:
Yeah, let's always add the parameter to the callback, provide undefined where the argument is omitted by the spec, and ignore is where unused by the spec within the callback body 👍

ADKaster commented:
Whoops, I re-requested a review from linus but it looks like this has conflicts with Fetch/Body.cpp now.

linusg commented:
@srikavin would you mind fixing the conflicts? :)

srikavin commented:
Whoops, I completely forgot about this.

I dropped the first two commits and rebased on master. I'm not entirely sure how to properly propagate errors relating to converting attribute values from `DeprecatedString` to `String`s with the current design, so I just used `release_value_but_fixme_should_propagate_errors`.

linusg commented on Userland/Libraries/LibWeb/HTML/HTMLLinkElement.cpp:
```diff
@@ -103,68 +104,313 @@ void HTMLLinkElement::parse_attribute(DeprecatedFlyString const& name, Deprecate
-            return;
-        }
+    // 7. Fetch request with processResponseConsumeBody set to the following steps given response response and null, failure, or a byte sequence bodyBytes:
+    (void)Fetch::Fetching::fetch(
```
This should probably not ignore errors, let's use `release_value_but_fixme_should_propagate_errors()` on the return value for better visibility

linusg reviewed (commented):
Looks great, thanks! Two small issues, happy to merge afterwards :)
